### PR TITLE
Implement IWantToRunWhenEndpointStarts

### LINKED ISSUE #60: Implement IWantToRunWhenEndpointStarts
Connects to https://github.com/Particular/NServiceBus/issues/3342

With `IWantToRunWhenBusStartsAndStops` becoming obsolete, each host will need to implement this functionality (including tests).

### PR DESCRIPTION
**Pull request by colin-higgins**
Closes #60 
Connects to https://github.com/Particular/NServiceBus/issues/3342

Bringing `IWantToRunWhenBusStartsAndStops` in as `IWantToRunWhenEndpointStartsAndStops`

### CODE REVIEW

**Comment by colin-higgins** [comment]
Removed WIP

@kbaley and @Particular/nservicebus-maintainers and @petersgiles 

Please review. Especially the split use of `IConfigureComponents` and `IBuilder`

**Comment by SeanFeldman** [comment]
@colin-higgins this change would need to go to https://github.com/Particular/NServiceBus.Host.AzureCloudService/issues/24. Too bad we can't keep it in one spot.

**Comment by colin-higgins** [comment]
@SeanFeldman That is next on the list! @kbaley and I will be pairing on that once this one is in.

**Comment by SeanFeldman** [comment]
I'd like to join when you do.

**Comment by petersgiles** [comment]
@colin-higgins Looks like you have a few redundant usings in the generichost

**Comment by colin-higgins** [comment]
@petersgiles I removed that using and also added some other cleanup.

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            }
+
+            public async Task Stop(IMessageSession session)
+            {
+                await Task.Yield();
+                throw new InvalidOperationException("AsyncThrowingStop");
+            }
+        }
+    }
+}
```
Comment (timbussmann) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: I think it would make sense to prefix the test names with the Method you're unit testing. (e.g. `Start_Should[...]`)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        public void Add(IWantToRunWhenEndpointStartsAndStops instance)
+        {
+            explicitInstances.Add(instance);
+        }
+
+        public IEnumerable<IWantToRunWhenEndpointStartsAndStops> Resolve(IConfigureComponents configurer, IBuilder builder)
+        {
+            foreach (var startAndStoppableType in scannedStartAndStoppableTypes)
+            {
+                configurer.ConfigureComponent(startAndStoppableType, DependencyLifecycle.InstancePerCall);
```
Comment (timbussmann) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs]: it's a weird behavior that `Resolve` also registers components in the container. From the interface this is not obvious and seems a SoC violation to me.

File: src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    <Reference Include="System.Data" />
+    <Reference Include="System.Net.Http" />
+    <Reference Include="System.Xml" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="When_starting_and_stopping_host.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="NServiceBus.Hosting.AcceptanceTests.snk" />
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj]: Do we strongname the acceptance tests?

File: src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    <None Include="packages.config" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\NServiceBus.Hosting.Windows\NServiceBus.Host.csproj">
+      <Project>{85E813C0-4A94-4946-8B1F-DE1E39AA7D11}</Project>
+      <Name>NServiceBus.Host</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <!-- To modify your build process, add your task inside one of the targets below and uncomment it.
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj]: Remove the comments

File: src/NServiceBus.Hosting.AcceptanceTests/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Reflection;
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.AcceptanceTests/Properties/AssemblyInfo.cs]: Delete the whole file?

File: src/NServiceBus.Hosting.AcceptanceTests/When_starting_and_stopping_host.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        class IWantToRunContext
+        {
+            public bool StartCalled { get; set; }
+            public bool StopCalled { get; set; }
+        }
+
+        class GenericEndpointConfig : IConfigureThisEndpoint
+        {
+            IWantToRunContext context;
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.AcceptanceTests/When_starting_and_stopping_host.cs]: Should be at the bottom of the file  (run a R# code cleanup with the "particular" conventions)

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+            var syncThrows = new SyncThrowingStart();
+            var thingsToBeStarted = new IWantToRunWhenEndpointStartsAndStops[] { startable1, startable2, syncThrows };
+
+            var runner = new StartableAndStoppableRunner(thingsToBeStarted);
+            try
+            {
+                await runner.Start(null);
+            }
+            // ReSharper disable once EmptyGeneralCatchClause
+            catch
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: Catch the specific `InvalidOperationException` thrown by the `SyncThrowingStart` ?

File: src/NServiceBus.Hosting.Windows/Arguments/HostArguments.cs [hunk 8]
```diff
@@ -9,8 +9,8 @@ namespace NServiceBus.Hosting.Windows.Arguments
 
     class HostArguments
     {
-        private readonly OptionSet installOptions;
-        private readonly OptionSet uninstallOptions;
+        readonly OptionSet installOptions;
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/Arguments/HostArguments.cs]: Remove readonly as well

File: src/NServiceBus.Hosting.Windows/GenericHost.cs [hunk 9]
```diff
@@ -88,6 +89,11 @@ Task<IStartableEndpoint> PerformConfiguration(Action<EndpointConfiguration> more
             RoleManager.TweakConfigurationBuilder(specifier, configuration);
             profileManager.ActivateProfileHandlers(configuration);
 
+
+            configuration.RegisterComponents(r => r.ConfigureComponent(() => startableAndStoppableBag, DependencyLifecycle.SingleInstance));
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/GenericHost.cs]: Think its better to move this into the setup method of the `StartableAndStoppableFeature` ?

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+                var startable1 = startable;
+                task.ContinueWith(t =>
+                {
+                    thingsRanAtStartup.Add(startable1);
+                    Log.DebugFormat("Started {0}.", startable1.GetType().AssemblyQualifiedName);
+                }, TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously);
+                task.ContinueWith(t => { Log.Error($"Startup task {startable1.GetType().AssemblyQualifiedName} failed to complete.", t.Exception); }, TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously);
+
+                startableTasks.Add(task);
```
Comment (timbussmann) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: why not just use async/await syntax instead of the TPL?

File: src/NServiceBus.Hosting.Windows/GenericHost.cs [hunk 11]
```diff
@@ -88,6 +89,11 @@ Task<IStartableEndpoint> PerformConfiguration(Action<EndpointConfiguration> more
...
+            configuration.RegisterComponents(r => r.ConfigureComponent<StartableAndStoppableTask>(DependencyLifecycle.InstancePerCall));
```
Comment (timbussmann) [inline src/NServiceBus.Hosting.Windows/GenericHost.cs]: can't we register these components in the feature setup?

File: src/NServiceBus.Hosting.Windows/GenericHost.cs [hunk 12]
```diff
@@ -39,6 +39,7 @@ public GenericHost(IConfigureThisEndpoint specifier, string[] args, List<Type> d
             }
 
             profileManager = new ProfileManager(assembliesToScan, args, defaultProfiles);
+            startableAndStoppableBag = new StartableAndStoppableBag(assembliesToScan);
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/GenericHost.cs]: If we move the registration to the feature (see below) you should be able to just iterate over the "scanned types" https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/EndpointConfiguration.cs#L207 ?

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
+﻿namespace NServiceBus
+{
+    using System.Threading.Tasks;
+    //using JetBrains.Annotations;
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs]: Remove comment

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+    /// <summary>
+    /// Implementers will be invoked when the endpoint starts up.
+    /// Dependency injection is provided for these types.
+    /// </summary>
+    //[UsedImplicitly(ImplicitUseTargetFlags.WithMembers)]
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs]: Remove comment

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs [hunk 15]
```diff
@@ -0,0 +1,55 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using ObjectBuilder;
+
+    class StartableAndStoppableBag : IHoldStartAndStoppables
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs]: Should we convert this to a feature startup task?

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        ConcurrentBag<IWantToRunWhenEndpointStartsAndStops> explicitInstances = new ConcurrentBag<IWantToRunWhenEndpointStartsAndStops>();
+        List<Type> scannedStartAndStoppableTypes;
+
+        public StartableAndStoppableBag(IEnumerable<Assembly> assemblies)
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs]: Just pass the types in from the feature setup method? (see comment above)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+        public StartableAndStoppableTask(IConfigureComponents configurer, IBuilder builder, IHoldStartAndStoppables startAndStoppablesBag)
+        {
+            var startAndStoppables = startAndStoppablesBag.Resolve(configurer, builder);
+            startAndStoppablesRunner = new StartableAndStoppableRunner(startAndStoppables);
+        }
+
+        protected override async Task OnStart(IMessageSession session)
+        {
+            await startAndStoppablesRunner.Start(session);
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: ConfigureAwait(false)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
+namespace NServiceBus
+{
+    using System.Threading.Tasks;
+    using Features;
+    using ObjectBuilder;
+
+    class StartableAndStoppableTask : FeatureStartupTask
+    {
+        StartableAndStoppableRunner startAndStoppablesRunner;
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: Code formatting (move to bottom of file)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        protected override async Task OnStart(IMessageSession session)
+        {
+            await startAndStoppablesRunner.Start(session);
+        }
+
+        protected override async Task OnStop(IMessageSession session)
+        {
+            await startAndStoppablesRunner.Stop(session);
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: ConfigureAwait(false)

**Comment by andreasohlund** [comment]
I've put the [WIP] flag back on.

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs (hunk 2 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs]: Yeah, I could rename have another method specifically for register. No problem there.

File: src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj (hunk 3 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj]: This was in order to sign and share internals

File: src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj (hunk 4 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj]: Will do

File: src/NServiceBus.Hosting.AcceptanceTests/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.AcceptanceTests/Properties/AssemblyInfo.cs]: Will do

File: src/NServiceBus.Hosting.AcceptanceTests/When_starting_and_stopping_host.cs (hunk 6 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.AcceptanceTests/When_starting_and_stopping_host.cs]: Will do

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs (hunk 7 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: Even though this is only a test? This was a copy paste from core.

File: src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj (hunk 3 above)
Comment (andreasohlund) [inline src/NServiceBus.Hosting.AcceptanceTests/NServiceBus.Hosting.AcceptanceTests.csproj]: Acceptance tests should only operate on the public api, which internals do you need to share?

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs (hunk 7 above)
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: Seems more safe?

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs (hunk 10 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: Done by me when I ported the original code. It was primarily done because of the conditional logging. Since we want to start them concurrently and log per instance there is not much we can do else. I initially asked if we could remove the logs. The answer was no.

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs (hunk 17 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: No @andreasohlund, the answer is remove `async` and return the task

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs (hunk 19 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: No @andreasohlund, the answer is remove `async` and return the task

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+{
+    using System.Threading.Tasks;
+    using Features;
+    using ObjectBuilder;
+
+    class StartableAndStoppableTask : FeatureStartupTask
+    {
+        StartableAndStoppableRunner startAndStoppablesRunner;
+
+        public StartableAndStoppableTask(IConfigureComponents configurer, IBuilder builder, IHoldStartAndStoppables startAndStoppablesBag)
```
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: Something with the dependencies is really off in this design. Now this class requires the configurer and the builder. I'll see if I can send in a proposal PR or work with @colin-higgins directly

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableBag.cs]: I agree with @timbussmann , see https://github.com/Particular/NServiceBus.Host/pull/64/files/95ce4d5aa0065c16e9d1a2f9891095c8c5ac32fa#r57150013

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs (hunk 19 above)
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: :+1:

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs (hunk 17 above)
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: Good catch, I bow to the master of async :)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs (hunk 14 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs]: This attribute is internal to core. I could copy paste it over here.........?

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs (hunk 20 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: This has been fixed in the latest commits

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs (hunk 7 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: Doesn't that just make the test more fragile? How is it more safe?

File: src/NServiceBus.Hosting.ComponentTests/When_starting_and_stopping_host.cs [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+            Assert.True(context.InstanceStartCalled);
+            Assert.True(context.InstanceStopCalled);
+        }
+
+        [Test]
+        public void Instance_is_registered_once()
+        {
+            var context = RunHost();
+
+            Assert.True(context.InstanceTimesRegistered == 1);
```
Comment (colin-higgins) [inline src/NServiceBus.Hosting.ComponentTests/When_starting_and_stopping_host.cs]: I added this test because I noticed that there was nothing excluding the registered instances from assembly scanning, so this is current behavior but it seems counter intuitive.

I don't know if there are any users that might be accidentally relying on this behavior?

Thoughts?

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs (hunk 7 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: Because it is more specific. I wrote that test. So blame me :dancer:

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: Well that is a matter of taste. We haven't really settled on conventions...

File: src/NServiceBus.Hosting.ComponentTests/When_starting_and_stopping_host.cs (hunk 21 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.ComponentTests/When_starting_and_stopping_host.cs]: I would move this out into seperate PR.

**Comment by danielmarbach** [comment]
:+1: like it.

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs (hunk 10 above)
Comment (kbaley) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: Is it worth adding a comment to this effect in the test?

**Comment by colin-higgins** [comment]
ping @andreasohlund and @timbussmann.
Ready for review

**Comment by danielmarbach** [comment]
Would it make sense to squash?

**Comment by colin-higgins** [comment]
Sure, I'll squash into one

File: src/NServiceBus.Hosting.Windows/GenericHost.cs [hunk 22]
```diff
@@ -88,6 +88,8 @@ Task<IStartableEndpoint> PerformConfiguration(Action<EndpointConfiguration> more
             RoleManager.TweakConfigurationBuilder(specifier, configuration);
             profileManager.ActivateProfileHandlers(configuration);
 
+            configuration.EnableFeature<StartableAndStoppableFeature>();
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/GenericHost.cs]: Just make the feature "enabled by default"? (in the ctor of the feature)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/EndpointConfigurationExtensions.cs [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+
+    /// <summary>
+    /// Extension methods for EndpointConfiguration
+    /// </summary>
+    public static class EndpointConfigurationExtensions
+    {
+        /// <summary>
+        /// Register a specific instance of an IWantToRunWhenEndpointStartsAndStops implementation
+        /// </summary>
+        public static void RunWhenEndpointStartsAndStops(this EndpointConfiguration configuration, IWantToRunWhenEndpointStartsAndStops callbackObject)
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/EndpointConfigurationExtensions.cs]: Is callbackObject a good name? (made me think about our callbacks)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/EndpointConfigurationExtensions.cs [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            var settings = configuration.GetSettings();
+
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/EndpointConfigurationExtensions.cs]: I think the code would be more clean if you do a `TryGet` instead? (and only do a `Set` if it wasn't found)

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+    /// Dependency injection is provided for these types.
+    /// </summary>
+    [UsedImplicitly(ImplicitUseTargetFlags.WithMembers)]
+    public interface IWantToRunWhenEndpointStartsAndStops
+    {
+        /// <summary>
+        /// Method called at startup.
+        /// </summary>
+        /// <exception cref="System.Exception">This exception will be thrown if <code>null</code> is returned. Return a Task or mark the method as <code>async</code>.</exception>
+        Task Start(IMessageSession session);
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs]: Add a api comment to explain what the session is?

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+                await Task.WhenAll(stoppableTasks.ToArray()).ConfigureAwait(false);
+            }
+            // ReSharper disable once EmptyGeneralCatchClause
+            catch
+            {
+                // ignore because we want to shutdown no matter what.
+            }
+        }
+
+        readonly IEnumerable<IWantToRunWhenEndpointStartsAndStops> wantToRunWhenBusStartsAndStops;
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: No need for readonly

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs [hunk 27]
```diff
@@ -0,0 +1,26 @@
+namespace NServiceBus
+{
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Features;
+
+    class StartableAndStoppableTask : FeatureStartupTask
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableTask.cs]: Nest the feature startup task class inside the feature class, more easy to see how it fits together then

File: src/NServiceBus.Hosting.ComponentTests/Properties/AssemblyInfo.cs [hunk 28]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Reflection;
+using System.Runtime.InteropServices;
+
+[assembly: AssemblyTitle("NServiceBus.Hosting.ComponentTests")]
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.ComponentTests/Properties/AssemblyInfo.cs]: Can we delete this file?

File: src/NServiceBus.Hosting.ComponentTests/When_starting_and_stopping_host.cs [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+                configuration.UseSerialization<JsonSerializer>();
+                configuration.EnableInstallers();
+                configuration.SendFailedMessagesTo("error");
+                configuration.UsePersistence<InMemoryPersistence>();
+                configuration.RegisterComponents(c => c.ConfigureComponent(() => context, DependencyLifecycle.SingleInstance));
+
+                configuration.RunWhenEndpointStartsAndStops(new RunStuffInstance(context));
+            }
+
+            internal class RunStuffInstance : IWantToRunWhenEndpointStartsAndStops
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.ComponentTests/When_starting_and_stopping_host.cs]: Can we remove the internal?

File: src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs (hunk 1 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs]: I renamed the tests.

**Comment by colin-higgins** [comment]
@Particular/nservicebus-maintainers PING

File: src/NServiceBus.Hosting.Windows/InstallWindowsHost.cs [hunk 30]
```diff
@@ -11,7 +11,7 @@ public class InstallWindowsHost : MarshalByRefObject
         GenericHost genericHost;
 
         /// <summary>
-        /// Accepts the type which will specify the users custom configuration.
+        /// Accepts the type which will specify the users custom configuration.B
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/InstallWindowsHost.cs]: ".B" ?

**Comment by andreasohlund** [comment]
looks good to me now.

@indualagarsamy are you able to take this for a spin to see that it aligns with what we discussed in Bangkok?

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+                }, TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously);
+                task.ContinueWith(t =>
+                {
+                    Log.Error($"Startup task {startable1.GetType().AssemblyQualifiedName} failed to complete.", t.Exception);
+                }, TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously);
+                
+                startableTasks.Add(task);
+            }
+
+            return Task.WhenAll(startableTasks.ToArray());
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: Since this blocks startup if users do a "while(true)" in their start runner should we add a timeout task to at least do some logging that "this startup runner takes a long time to start"? (cc @indualagarsamy )

Also we need to spell this out very clear in API doco that user should not do any blocking stuff in there. I v5 this was ok since we allowed ".Start" to never complete and users porting their code to v6 will hit this and their endpoints will hang on startup.

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+
+    /// <summary>
+    /// Implementers will be invoked when the endpoint starts up.
+    /// Dependency injection is provided for these types.
+    /// </summary>
+    [UsedImplicitly(ImplicitUseTargetFlags.WithMembers)]
+    public interface IWantToRunWhenEndpointStartsAndStops
+    {
+        /// <summary>
+        /// Method called at startup.
```
Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs]: Should we spell out that users should not do blocking operations here? (since in v5 they could without any bad effects - http://docs.particular.net/nservicebus/lifecycle/iwanttorunwhenbusstartsandstops#versions-5-and-below )

File: src/NServiceBus.Hosting.Windows/InstallWindowsHost.cs (hunk 30 above)
Comment (colin-higgins) [inline src/NServiceBus.Hosting.Windows/InstallWindowsHost.cs]: Ugh. Must have messed up a hot key combination

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs (hunk 31 above)
Comment (timbussmann) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: I think the Task return value indicates that we expect the Start method to finish. The docs already state:

> The call to IStartableEndpoint.Start() will not return before all instances of IWantToRunWhenBusStartsAndStops.Start() are completed.

I'd say everything else is feature creep. How about raising an issue about some additional logging?

Comment (andreasohlund) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: Me and @indualagarsamy wasted a good hour in bangkok on this mistake and it was the main driver for moving the interface to the host  so I'd say it's in scope?

**Comment by timbussmann** [comment]
I think the build needs to be adjusted to run `NServiceBus.Hosting.ComponentTests` in case they exist?

**Comment by colin-higgins** [comment]
Yes, I'll set those up. I just wanted to make sure this pull request was good to go first.

**Comment by timbussmann** [comment]
@colin-higgins merged

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs (hunk 31 above)
Comment (indualagarsamy) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: @colin-higgins @timbussmann - Did @andreasohlund's suggestion make it into the PR? Yes, it's important for us to have this WARNING log to let the users know that their endpoint hasn't started, if we haven't done so.

Comment (indualagarsamy) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: @timbussmann - why was this PR merged? When there was a discussion about an important aspect in discussion?

Comment (timbussmann) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: @indualagarsamy - @andreasohlund and I discussed this on a zoom call and we agreed that it makes sense to do, but that it's also a different PR since this is about moving the existing behavior from the core to the host. Seems like we forgot the update the discussion here.

Comment (bording) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: @indualagarsamy For reference, this is being added by #72.

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs (hunk 32 above)
Comment (danielmarbach) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs]: In Bangkok we concluded that renaming the interface should be enough to guide the users since it is a new interface and we educate our customers what it means to return a task.

File: src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs (hunk 31 above)
Comment (indualagarsamy) [inline src/NServiceBus.Hosting.Windows/StartableAndStoppable/StartableAndStoppableRunner.cs]: @bording - I understand. It's just that I wasn't aware that there was a discussion and that it was decided to do it as a separate PR.
